### PR TITLE
17941 add agm_location_change_schema

### PR DESCRIPTION
**Pull request by kzdev420**
*Issue #:* /bcgov/entity#17941

*Description of changes:*

Add AGM_Location_change Schema

- existing deliveryInstructions field in address schema will be used for additionalLocationDetails
	
- required addressRegion validation will be done in the legal api validators to avoid needing to overriding existing address schema validation

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-schemas license (Apache 2.0).

### CODE REVIEW

File: src/registry_schemas/schemas/agm_location_change.json [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                            "maxLength": 40
+                        },
+                        "addressCountry": {
+                            "type": "string",
+                            "enum": ["Canada", "CA"]
+                        },
+                        "addressRegion": {
+                            "type": "string",
+                            "enum": ["BC"],
+                            "description": "Must be in B.C."
```
Comment (vysakh-menon-aot) [inline src/registry_schemas/schemas/agm_location_change.json:55]: Should not be in BC. Yet to update design

Comment (vysakh-menon-aot) [reply src/registry_schemas/schemas/agm_location_change.json:55]: Probably you can skip this validation in schema. I will add this in legal-api

Comment (kzdev420) [reply src/registry_schemas/schemas/agm_location_change.json:53]: I removed that

File: src/registry_schemas/schemas/agm_location_change.json [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                        },
+                        "addressCity": {
+                            "type": "string",
+                            "maxLength": 40
+                        },
+                        "addressCountry": {
+                            "type": "string"
+                        },
+                        "addressRegion": {
+                            "type": "string",
```
Comment (vysakh-menon-aot) [inline src/registry_schemas/schemas/agm_location_change.json:52]: Can you add max length? refer: https://github.com/bcgov/business-schemas/blob/main/src/registry_schemas/schemas/address.json#L31

Comment (kzdev420) [reply src/registry_schemas/schemas/agm_location_change.json]: I added it

File: src/registry_schemas/schemas/agm_location_change.json [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                "newAgmLocation"
+            ],
+            "properties": {
+                "year": {
+                    "type": "string",
+                    "description": "Year of the AGM, Must be on or after incorporation year and cannot be future year."
+                },
+                "newAgmLocation": {
+                    "type": "object",
+                    "description": "New AGM Location, Must be in B.C.",
```
Comment (vysakh-menon-aot) [inline src/registry_schemas/schemas/agm_location_change.json:24]: must not be in BC

Comment (kzdev420) [reply src/registry_schemas/schemas/agm_location_change.json]: I updated

**Review by vysakh-menon-aot — APPROVED** [review]

File: tests/unit/test_agm_location_change.py
Comment (argush3) [inline tests/unit/test_agm_location_change.py:1]: Can we also add a test that shows that validation fails if `newAgmLocation` is not provided?

Comment (kzdev420) [reply tests/unit/test_agm_location_change.py:1]: I added that case

**Review by argush3 — APPROVED** [review]

File: src/registry_schemas/schemas/agm_location_change.json [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            "required": [
+                "year",
+                "newAgmLocation"
+            ],
+            "properties": {
+                "year": {
+                    "type": "string",
+                    "description": "Year of the AGM, Must be on or after incorporation year and cannot be future year."
+                },
+                "newAgmLocation": {
```
Comment (thorwolpert) [inline src/registry_schemas/schemas/agm_location_change.json:22]: Could this use the Address Schema?

File: src/registry_schemas/example_data/schema_data.py [hunk 5]
```diff
@@ -2701,6 +2701,19 @@
...
 
+AGM_LOCATION_CHANGE = {
+    'year': '2023',
+    'newAgmLocation': {
+        'streetAddress': 'mailing_address - address line one',
+        'streetAddressAdditional': '',
+        'addressCity': 'mailing_address city',
+        'addressCountry': 'CA',
+        'postalCode': 'H0H0H0',
+        'addressRegion': 'ON',
```
Comment (thorwolpert) [inline src/registry_schemas/example_data/schema_data.py:2712]: Is this allowed to be any country & province/state?
I ask as you example is not in BC

**Review by thorwolpert — COMMENTED** [review]
Mostly a question about using the Address Schema?

File: src/registry_schemas/schemas/agm_location_change.json (hunk 4 above)
Comment (argush3) [reply src/registry_schemas/schemas/agm_location_change.json:22]: Actually, yes I think we can.  

Maybe something along the lines of this.  Filing schema does something similar too -> https://github.com/bcgov/business-schemas/blob/main/src/registry_schemas/schemas/filing.json#L7-L23

Does something like this work for you, @thorwolpert ?

```json
"newAgmLocation": {
      "allOf": [
        { "$ref": "https://bcrs.gov.bc.ca/.well_known/schemas/address" },
        {
          "properties": {
				"additionalLocationDetails": {
                 	"type": ["string", "null"],
                    "maxLength": 2000 }
          },
        }]
}
```

File: src/registry_schemas/example_data/schema_data.py (hunk 5 above)
Comment (argush3) [reply src/registry_schemas/example_data/schema_data.py:2712]: The `addressRegion` can't be BC.  Vysakh will be adding validation in the legal api for this.

File: src/registry_schemas/schemas/agm_location_change.json (hunk 4 above)
Comment (kzdev420) [reply src/registry_schemas/schemas/agm_location_change.json:22]: I think it will work

Comment (kzdev420) [reply src/registry_schemas/schemas/agm_location_change.json:22]: Hi @thorwolpert 

I updated the `newAgmLocation` to use the Address Schema.
Please review it again.

File: src/registry_schemas/schemas/agm_location_change.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                    "allOf": [
+                        { "$ref": "https://bcrs.gov.bc.ca/.well_known/schemas/address" },
+                        {
+                            "properties": {
+                                "additionalLocationDetails": {
```
Comment (kialj876) [inline src/registry_schemas/schemas/agm_location_change.json:27]: Is there a difference between this and 'streetAddressAdditional' used in the address schema?

Comment (kzdev420) [reply src/registry_schemas/schemas/agm_location_change.json:27]: Yes, they are different

File: src/registry_schemas/schemas/agm_location_change.json [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                                    "type": ["string", "null"],
+                                    "maxLength": 2000
+                                }
+                            },
+                            "required": [
```
Comment (kialj876) [inline src/registry_schemas/schemas/agm_location_change.json:32]: Shouldn't these already be required/verified via the address schema?

File: src/registry_schemas/example_data/schema_data.py [hunk 8]
```diff
@@ -2701,6 +2701,19 @@
...
+        'additionalLocationDetails': ''
```
Comment (argush3) [inline src/registry_schemas/example_data/schema_data.py:2713]: can we update the sample data to use the delivery instructions field?

Comment (kzdev420) [reply src/registry_schemas/example_data/schema_data.py]: I updated

**Review by argush3 — APPROVED** [review]

**Review by argush3 — APPROVED** [review]

File: src/registry_schemas/schemas/agm_location_change.json (hunk 6 above)
Comment (kialj876) [reply src/registry_schemas/schemas/agm_location_change.json]: **delivery instructions -- you updated it 👍

**Review by kialj876 — APPROVED** [review]
Thank you for the changes 👍

**Pull request merged**